Src/PerformanceCollector/Filtering/CollectionConfiguration.cs

```diff
@@ -0,0 +1,331 @@
+
+            try
+            {
+                metrics.Add(new OperationalizedMetric<TTelemetry>(metricInfo, out errors));
+            }
+            catch (Exception e)
+            {
+                // error creating the metric
+                errors =
+                    errors.Concat(
+                        new[]
+                        {
+                            CollectionConfigurationError.CreateError(
+                                CollectionConfigurationErrorType.MetricFailureToCreate,
+                                string.Format(CultureInfo.InvariantCulture, "Failed to create metric {0}.", metricInfo),
+                                e,
+                                Tuple.Create("MetricId", metricInfo.Id))
+                        }).ToArray();
+            }
+        }
+
+        private void CreatePerformanceCounters(out CollectionConfigurationError[] errors)
+        {
+            var errorList = new List<CollectionConfigurationError>();
+
+            OperationalizedMetricInfo[] performanceCounterMetrics =
+                (this.info.Metrics ?? new OperationalizedMetricInfo[0]).Where(metric => metric.TelemetryType == TelemetryType.PerformanceCounter)
+                    .ToArray();
+
+            this.performanceCounters.AddRange(
+                performanceCounterMetrics.GroupBy(metric => metric.Id, StringComparer.Ordinal)
+                    .Select(group => group.First())
+                    .Select(pc => Tuple.Create(pc.Id, pc.Projection)));
+
+            IEnumerable<string> duplicateMetricIds =
+                performanceCounterMetrics.GroupBy(pc => pc.Id, StringComparer.Ordinal).Where(group => group.Count() > 1).Select(group => group.Key);
+
+            foreach (var duplicateMetricId in duplicateMetricIds)
+            {
+                errorList.Add(
+                    CollectionConfigurationError.CreateError(
+                        CollectionConfigurationErrorType.PerformanceCounterDuplicateIds,
+                        string.Format(CultureInfo.InvariantCulture, "Duplicate performance counter id '{0}'", duplicateMetricId),
+                        null,
+                        Tuple.Create("MetricId", duplicateMetricId)));
+            }
+
+            errors = errorList.ToArray();
+        }
+
+        private void CreateDocumentStreams(
+            out CollectionConfigurationError[] errors,
+            Clock timeProvider,
+            IEnumerable<DocumentStream> previousDocumentStreams)
+        {
+            var errorList = new List<CollectionConfigurationError>();
+            var documentStreamIds = new HashSet<string>();
+
+            // quota might be changing concurrently on the collection thread, but we don't need the exact value at any given time
+            // we will try to carry over the last known values to this new configuration
+            Dictionary<string, Tuple<float, float, float, float, float>> previousQuotasByStreamId =
```

Comment:
There's a global quota (which is not part of the configuration), and also individual per-telemetry type quotas for each document stream. They have to be independent for each stream to ensure that a chatty stream doesn't use up the entire quota and other streams (and users) get nothing.